Add beginner friendly documentation (#102)
mobxjs/mobx-react-docz: content/basics/basics.mdx

---

> name: Basics

Reviewer: What about "Getting started"? :)

route: /basics
---

import { Playground } from 'docz'

# Basic Introduction

> [MobX](https://mobx.js.org/README.html) lets you create observable state. This

Reviewer: Link to the site directly in case we would restructure it in the future.
```suggestion
[MobX](https://mobx.js.org) lets you create observable state. This
```

library, `mobx-react` integrates it with React. It provides a way for you to

> create React components that respond to MobX state changes.

Reviewer: The library isn't really creating anything, just wrapping.
```suggestion
enhance React components that respond to MobX state changes.
```

Author: I've rewritten it to "a way for you to make React components respond to MobX state changes now" now.

> Frameworks such as [mobx-state-tree](https://mobx-state-tree.js.org) and
> [mobx-keystone](https://mobx-keystone.js.org/) build on
> top of MobX to offer a higher-level way to create observable models. These models
> give you additional properties, such as automatic serialization to JSON. This
> introduction uses plain MobX models. The principles of React

> > integration are the same for these advanced libraries are the same

Reviewer: This ending of this sentence doesn't make much sense, try to rewrite it differently.

> as they are also built with MobX.

## MobX models

Let's consider MobX models for animals in a zoo, where each animal
has a certain amount of energy. If the energy is below 50, the animal
is hungry. Animals can be fed to increase their energy levels:

> These code examples use TypeScript. If you use plain JavaScript you
> can skip the type declarations.

```tsx
import { observable, action, computed } from 'mobx'

class Animal {
  name: string
  species: string

>   @observable energy: number

Reviewer: Given that upcoming MobX 6 is dropping support for decorators, let's not spread it here. Check https://github.com/mobxjs/mobx/issues/2325 for API examples that are most likely set in the stone by now.

Author: Okay, I didn't know that and I need to look into that, thanks!

Reviewer: Continuing here with [your concerns about decorators](https://github.com/mobxjs/mobx-react-docz/pull/102#issuecomment-646591780). I am not sure what is your confusion. The first comment shows "API changes" section on how it will look like with MobX 6. I know it's not technically available yet, but there is plenty of examples with decorators in other docs. Here I would like to target it more toward the future. We can even include some warning this API is not available yet. We can have it here like "promotion" on what's will become "next standard"

Author: I am not interested in documenting an API that isn't available yet.

  constructor(name: string, species: string, energy: number = 100) {
    this.name = name
    this.species = species
    this.energy = energy
  }

  @action
  timePasses() {
    this.energy -= 10
  }

  @action
  feed() {
    this.energy += 50
    if (this.energy > 100) {
      this.energy = 100
    }
  }

  @computed
  get isHungry() {
    return this.energy < 50
  }
}

class Zoo {
  @observable animals: Animal[] = []

  @action
  addAnimal(animal: Animal) {
    this.animals.push(animal)
  }
}
```

## A React component that observes

We haven't used anything except plain MobX in this example. Now let's
create a React component to show information for an animal:

```tsx
import { Component } from 'react'
import { observer } from 'mobx-react'

> const AnimalView = observer(({ animal }: { animal: Animal }) => {

Reviewer: It's a matter of preference, but I think this is more readable as there is less animals together :)
```suggestion
type TProps = {
  animal: Animal
}
const AnimalView = observer<TProps>(({ animal }) => {
```

    </div>
  )
})
```

> These examples work with `mobx-react-lite` too. `mobx-react` expands
> `mobx-react-lite` to have support for React class components. If you are
> creating a new project that only uses React function components, you can use
> `mobx-react-lite`.

Let's create an instance of `Animal`:

```tsx
const fred = new Animal('Fred', 'hippo')
```

We can now render `fred`:

```tsx
<AnimalView animal={fred} />
```

The only special thing that `AnimalView` does compared to a normal React
function component is that it is wrapped in `observer` from `mobx-react`. This
makes `AnimalView` automatically re-render itself if anything that it observes
changes. In this case the observable information is the computed property
`isHungry`, which in turn it based on the observable property `energy`.
`AnimalView` automatically re-renders itself only if `isHungry` changes. You
don't need to do anything else to make that happen.

When you wrap your React component in `observer`, it tracks any observable
property or object you access during rendering it, and reacts to any changes to
these. It won't react to any changes in observable properties you _don't_

> directly or indirectly use during rendering.

Reviewer: What would "indirectly" mean? Perhaps add an example for that?

## When do observers observe?

Note that MobX can only observe properties, not simple values like a `string`
or a `number`. So if you write your component to only receive simple values, it
does not benefit from observability:

```tsx
import { Component } from 'react'
import { observer } from 'mobx-react'

> // Uses observer() but does not benefit from observability

Reviewer: Have you tried it? I believe that destructuring props like is still fine because `observer` sees access to those properties.

const NonObservingAnimalView = observer(
  ({
    name,
    species,
    isHungry,
  }: {

>     name: string

Reviewer: Same here as above, I would separate TProps to make it more readable.

    species: string
    isHungry: boolean
  }) => {
    return (
      <div>
        {name} is of species {species} and is{' '}
        {isHungry ? 'hungry' : 'not hungry'}.
      </div>
    )
  },
)
```

The component `NonObservingAnimalView` works perfectly if it is used within
another component that receives an `Animal` instance and is also an observer,
but it re-render each and every time, because MobX cannot see whether
`name`, `species` or `isHungry` are changed. Our original `AnimalView` _does_
observe it, because it can track property accesses.

> Wrapping a component in `observer` is cheap, so if you use MobX with react,

Reviewer: ```suggestion
Wrapping a component in `observer` is cheap, so if you use MobX with React,
```

> it's recommended to just wrap all your components in `observer`, even if the

Reviewer: I wouldn't say it's "recommended". It's an escape hatch so you don't need to think about it. The recommended is to use https://mobx.js.org/refguide/api.html#computedrequiresreaction-boolean

component does not use any observable information yet. There is no need
distinguish between React components that observe and do not observe in this.

## Observable arrays

> MobX does support observable arrays, observable
> objects and observable maps, so those _can_ be observed for changes by
> a React component. We demonstrate that here with `animals`:

Reviewer: ```suggestion
MobX does support arrays, objects, Maps, and Sets. Those _can_ be observed for changes by a React component. We demonstrate that here with `animals`:
```

```tsx
import { Component } from 'react'
import { observer } from 'mobx-react'

const AnimalsView = observer(({ animals }: { animals: Animal[] }) => {
  return (
    <ul>
      {animals.map(animal => (
        <ul>
          <AnimalView key={animal.name} animal={animal} />
        </ul>
      ))}
    </ul>
  )
})
```

Instances of our `Zoo` class have an `animals` property that is observable, so
let's use that:

```tsx
const wilma = new Animal('Wilma', 'penguin')

const zoo = new Zoo()
zoo.addAnimal(fred)
zoo.addAnimal(wilma)
```

We can now render our animals:

```tsx
<AnimalsView animals={zoo.animals} />
```

> `AnimalsView` automatically re-renders itself whenever you change the array, in

Reviewer: It would be worth to mention common pitfall here where people add `observer` around parent component rendering `AnimalsView` but to that component alone. In such a case, adding/removing from such array won't be reactive.

this case by using `Zoo.addAnimal`. The individual `AnimalView` re-renders
itself as before when any observed properties the underlying object change,
which in this case is its `isHungry` status.

## Local component state with `useLocalStore`

So far we have discussed observable model state, where the models live outside
of the React components. It is useful to have component-specific state however.

> This is state that only makes sense within that component, is typically very UI
> specific, and there is no reason to specify this part of your application's
> models.

Reviewer: ```suggestion
This is a state that only makes sense within that component. It is typically very UI
specific, and there is no reason to specify this part of your application's
models.
```

> This is useful for "temporary data", such as for instance the text that a user

Reviewer: I don't like the way you are treating `useLocalStore` here. It's far from "temporary data" only. It's a regular observable you can pass around, including actions & computeds. You can even stick into the Context and use anywhere in the app. It's equivalent to class-based stores. I know, the name of that hook tells something else, but it's too late to change that now.

Author: This is a tricky issue. You're right of course in that you can do anything with useLocalState. But I want to make sure the documentation doesn't offer too many options at once, and highlights the model-driven nature of many MobX applications. And I do want to mention the use case of React component specific state. I've added a bit of language to make clear it can do more for you too.

Reviewer: I agree. As I said in a comment below, we shouldn't mention `useLocalStore` here at all. Feel free to extract this portion to a separate file that won't be linked anywhere and we will see how it ends up.

Author: This means that a very commonly used pattern with MobX has no equivalent with functional components. I mean this one:

```js
class MyReactComponent extends Component {
  @observable isOpen = false
}
```
 
Are functional components with hooks read for prime time?

    local.name = ''
    local.species = ''
  }

  const handleChangeName = (e: React.ChangeEvent<HTMLInputElement>) => {
    local.name = e.target.value
  }

  const handleChangeSpecies = (e: React.ChangeEvent<HTMLInputElement>) => {
    local.species = e.target.value
  }

  return (
    <div>
      <div>
        Name:
        <input type="text" value={local.name} onChange={handleChangeName} />
      </div>
      <div>
        Species:
        <input
          type="text"
          value={local.species}
          onChange={handleChangeSpecies}
        />
      </div>
      <button onClick={handleAdd}>Add animal</button>
    </div>
  )
})
```

Here we define an object called `local` with two observable properties, `name`
and `species`. We use the React [controlled
components](https://reactjs.org/docs/forms.html#controlled-components)
technique to allow the user to edit these properties. As you can see, this
simply involves mutating the local state directly. We also have an `Add animal`
button which uses the local state to modify the zoo using the `addAnimal`
action. It then needs to reset the local state again to allow the user to
enter another animal.

> XXX interaction with strict mode, see https://github.com/mobxjs/mobx-react-lite/issues/235

Reviewer: XXX? :)

Author: Yes, a "TODO" comment. The problem is that this whole example breaks down, if I understand it correctly, as soon as you use MobX strict mode. On this issue you suggest useLocalStore is not worth it in strict mode. But what then would be the equivalent for the local component state use case? 

I want to write a document with *one* way to do it, and I either have to go to add text that says you shouldn't use MobX strict mode, or text that says, use MobX strict mode and this is what you use (but what? not useLocalStore).

Author: In addition I just read that MobX 6 wants to make 'observed' be the default. So if that doesn't go together with `useLocalStore` we need to go back to the drawing board completely for this section.

Reviewer: Yes, you are right. I think we will need to rethink `useLocalStore` anyway with MobX 6, it won't be that useful anymore. I will open a new issue to track that.

Reviewer: Actually, we are already tracking this problem in https://github.com/mobxjs/mobx-react-lite/issues/259. It would be probably best to remove `useLocalStore` from this document completely because it's uncertain how it will look like in the end.

if strict mode is the recommended story, what would we recommend instead of
use useLocalStore?

## Application organization

In these examples we described how you can create observer React components
that take observable instances as props, and this is a good pattern to follow.

But where do you get these props from? Often they can be retrieved from
_other_ observable objects. Let's consider `ZooView`:

```tsx
import { Component } from 'react'
import { observer } from 'mobx-react'

const ZooView = observer(({ zoo }: { zoo: Zoo }) => {
  return (
    <div>
      <AnimalsView animals={zoo.animals} />
      <AddAnimalView zoo={zoo} />
    </div>
  )
})
```

Where does `AnimalsView` get its `animals` prop from? From the `zoo` in
`ZooView`. Where does `AnimalView` get its `animal` prop? From the each animal
in the `AnimalsView`. We also pass `zoo` into `AddAnimalView` directly.

But where would we put the `zoo`? One simple approach is to make it a global in
your application, and simply import it when you need it.

Often this is sufficient. Consider a `Locale` MobX model that describes information

> about your site's Locale. You can simply maintain the current locale as

Reviewer: Let's warn here that global variables might be problematic for integration testing, it would cause flakiness of such tests.

Author: I think this section went against my own philosophy of "one way to do it" for this document by offering a lot of competing options, so I cut out the discussion of Context for now. I eventually want to add a section called "Context and cross-cutting concerns" but that can wait.

Reviewer: Yea, context is certainly more advanced, and in simple cases passing through props works justice fine. But I would like to have here a clear notion that stores in global variables might bite you.
